// @yakumoran/core/dist/core/plugin/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Decorators = void 0;
class Decorators {
    constructor() {
        this.commands = [];
        this.middlewares = new Map();
        this.events = new Map();
    }
    /**
     * @description 注册命令(需要返回一个boolean值，表示是否继续执行后续命令和中间件)
     * @param command 命令选项
     */
    Command(command) {
        return (target, propertyKey, descriptor) => {
            this.commands.push({
                options: command,
                handle: descriptor.value,
                _this: target
            });
        };
    }
    /**
     * @description 注册中间件
     * @param event 事件名
     * @param inBottom 是否在最后执行
     */
    Middleware(event, inBottom = false) {
        return (target, propertyKey, descriptor) => {
            if (!this.middlewares.has(event))
                this.middlewares.set(event, []);
            const list = this.middlewares.get(event);
            list.push({
                event,
                handle: descriptor.value,
                inBottom,
                _this: target
            });
            this.middlewares.set(event, list);
        };
    }
    /**
     * @description 注册事件监听器
     * @param event 事件名
     */
    EventListener(event, robot = false) {
        return (target, propertyKey, descriptor) => {
            if (!this.events.has(event))
                this.events.set(event, []);
            const list = this.events.get(event);
            list.push({
                event,
                handle: descriptor.value,
                _this: target,
                robot
            });
            this.events.set(event, list);
        };
    }
}
exports.Decorators = Decorators;
