// @yakumoran/core/dist/core/web/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebServer = void 0;
// @ts-ignore
const express_1 = __importDefault(require("express"));
const path_1 = require("path");
const logger_1 = require("../logger");
class WebServer {
    constructor(port) {
        this.logger = new logger_1.Logger('WebServer');
        this.app = (0, express_1.default)();
        this.port = port;
        this.app.use(express_1.default.json());
        this.app.use(express_1.default.urlencoded({ extended: true }));
        this.app.use(express_1.default.static((0, path_1.join)(__dirname, 'public')));
        this.start();
    }
    start() {
        this.app.listen(this.port, () => {
            this.logger.info(`WebServer 启动成功，监听端口 ${this.port}`);
        });
    }
    routeList() {
        return this.app._router.stack
            .filter((r) => r.route)
            .map((r) => {
            return {
                path: Object.keys(r.route.methods)[0].toUpperCase() + ' ' + r.route.path
            };
        });
    }
    route(path, router) {
        this.app.use(path, router);
    }
    unroute(path) {
        this.app._router.stack = this.app._router.stack.filter((r) => {
            if (r.name === 'router' && r.regexp.test(path))
                return false;
            return true;
        });
    }
}
exports.WebServer = WebServer;
