// @yakumoran/core/dist/core/packet/decoder/MailboxMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const html_entities_1 = require("html-entities");
exports.default = (message) => {
    if (/^@/.test(message)) {
        message.slice(2).split('<').forEach(e => {
            const tmp = e.split('>');
            if (tmp.length === 3) {
                return [[
                        'RoomNotice',
                        {
                            notice: (0, html_entities_1.decode)(tmp[0]),
                            background: tmp[1],
                            timestamp: Number(tmp[2])
                        }
                    ]];
            }
            if (tmp.length === 7) {
                if (/^'\^/.test(tmp[3])) {
                    const data = {
                        username: (0, html_entities_1.decode)(tmp[0]),
                        avatar: tmp[1],
                        gender: tmp[2],
                        background: tmp[4],
                        timestamp: Number(tmp[5]),
                        color: tmp[6]
                    };
                    return [['Follower', data]];
                }
                else if (/^'\*/.test(tmp[3])) {
                    const data = {
                        username: (0, html_entities_1.decode)(tmp[0]),
                        avatar: tmp[1],
                        gender: tmp[2],
                        background: tmp[4],
                        timestamp: Number(tmp[5]),
                        color: tmp[6],
                        message: (0, html_entities_1.decode)(tmp[3].substr(3))
                    };
                    return [['Like', data]];
                }
                else if (/^'\$/.test(tmp[3])) {
                    const data = {
                        username: (0, html_entities_1.decode)(tmp[0]),
                        avatar: tmp[1],
                        gender: tmp[2],
                        money: parseInt(tmp[3].split(' ')[0].substr(2)),
                        message: (0, html_entities_1.decode)(tmp[3].split(' ')[1] || ''),
                        background: tmp[4],
                        timestamp: Number(tmp[5]),
                        color: tmp[6]
                    };
                    return [['Payment', data]];
                }
            }
        });
    }
    return undefined;
};
